src/EditorFeatures/Core/CommandHandlers/AbstractGoToCommandHandler.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.FindUsages;
+using Microsoft.CodeAnalysis.Editor.Host;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.FindUsages;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.Notification;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Commanding;
+using Microsoft.VisualStudio.Text.Editor.Commanding;
+using VSCommanding = Microsoft.VisualStudio.Commanding;
+
+namespace Microsoft.CodeAnalysis.Editor.CommandHandlers
+{
+    internal abstract class AbstractGoToCommandHandler<TLanguageService, TCommandArgs> : VSCommanding.ICommandHandler<TCommandArgs>
+        where TLanguageService : class, ILanguageService
+        where TCommandArgs : EditorCommandArgs
+    {
+        private readonly IStreamingFindUsagesPresenter _streamingPresenter;
+        private readonly IThreadingContext _threadingContext;
+
+        public abstract string DisplayName { get; }
+
+        protected abstract string _scopeDescription { get; }
+
+        protected abstract FunctionId _functionId { get; }
+
+        protected abstract Task FindAction(TLanguageService service, Document document, int caretPosition, IFindUsagesContext context);
+
+        public AbstractGoToCommandHandler(
+            IThreadingContext threadingContext,
+            IStreamingFindUsagesPresenter streamingPresenter)
+        {
+            _threadingContext = threadingContext;
```

Comment:
Is it used anywhere?

Reply:
Used below in `GetFullyLoadedOpenDocumentInCurrentContextWithChanges`.

Reply:
Thank you, @olegtk !
Yes, it is used here: https://github.com/dotnet/roslyn/pull/35779/files#diff-a2844180b09d9705636953e332eb7bddR71 to check that we are not in the UI thread.